shared_code/utils.py

```diff
@@ -145,3 +151,50 @@ def locate_potential_duplicate(name, zipcode, client, collection):
     if dupe_candidate is not None:
         return dupe_candidate["name"]
     return False
+
+
+def validate_data(df):
+    # Check if all the required columns ('name', 'address1', 'city', 'state', 'zip', 'serviceSummary')
+    # are in the data frame
+    requiredColumns = ['name', 'address1', 'city', 'state', 'zip', 'serviceSummary']
+    missingColumns = []
+    for column in requiredColumns:
+        if not column in df.columns:
+            missingColumns.append(column)
+    if len(missingColumns) > 0:
+        raise Exception('The data frame does not contain the following required column(s): ' + str(missingColumns))
+
+    columns_to_display = ['name', 'city', 'state', 'zip']
+
+    found_error = False
+
+    # Identifies any row that contains a null or 'NONE' value in the 'name', 'address1', or 'serviceSummary' columns as
+    # invalid
+    for column in ['name', 'address1', 'serviceSummary']:
+        df_errors = df[empty_slots(df, column)]
+        if len(df_errors) > 0:
+            found_error = True
+            logger.error(" null or invalid values found for \'" + column + "\' column in rows: "
+                         + str(df_errors.index.values))
+            logger.error(df_errors[columns_to_display])
+
+    # Identifies any row that contains a null or 'NONE' value in the 'city', 'state', and 'zip' columns as invalid
+    df_errors = df[(empty_slots(df, 'city') | empty_slots(df, 'state')) & empty_slots(df, 'zip')]
```

Comment:
Hmmm not sure about the logic here, if zip is empty but city and state values are both provided, then it it's a good address isn't it.  You can pretty much locate the resource with `address, city, state` info or `address, zip` info.

Reply:
If `city` and `state` are provided but `zip` is empty, then we will have `(False OR False) AND True`, which will evaluate to `False` and therefore not get added to `df_errors`.

Here is a table that describes the logical outcomes of all possible cases, where `True` means it is absent in the row or is added to `df_errors` while `False` means it is present in the row or it is not added to `df_errors`:
|City|State|Zip|Logic Statement|Outcome|
|-|-|-|-|-|
|Present|Present|Present|(False OR False) AND False|False|
|Present|Present|Absent|(False OR False) AND True|False|
|Present|Absent|Present|(False OR True) AND False|False|
|Absent|Present|Present|(True OR False) AND False|False|
|Present|Absent|Absent|(False OR True) AND True|True|
|Absent|Present|Absent|(True OR False) AND True|True|
|Absent|Absent|Present|(True OR True) AND False|False|
|Absent|Absent|Absent|(True OR True) AND True|True|

The conclusion is that the row will only be inserted into `df_errors` if `zip` is absent AND if at least one of `city` and `state` is also absent.